src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestTPFatTPFServerWithTranslation.java

```diff
@@ -0,0 +1,236 @@
+			}
+		}
+		return unionTranslation;
+	}
+	
+	protected Set<Triple> handleGroupPattern(final SPARQLGroupPattern gp, final FederationAccessManager fedAccessMgr) throws FederationAccessException{
+		Set<Triple> groupTranslation = null;
+		for(final SPARQLGraphPattern i : gp.getSubPatterns()) {
+			if (i instanceof TriplePattern) {
+				if (groupTranslation == null) {
+					groupTranslation = handleTriplePattern((TriplePattern) i, fedAccessMgr);
+				} else {
+					final Set<Triple> partialTranslation = new HashSet<>();
+					for(final Triple j : handleTriplePattern((TriplePattern) i, fedAccessMgr)) {
+						if (groupTranslation.contains(j)) {
+							partialTranslation.add(j);
+						}
+					}
+					groupTranslation = partialTranslation;
+				}
+			} else if (i instanceof SPARQLUnionPattern) {
+				if (groupTranslation == null) {
+					groupTranslation = handleUnionPattern(((SPARQLUnionPattern) i), fedAccessMgr);
+				} else {
+					final Set<Triple> partialTranslation = new HashSet<>();
+					for(final Triple j : handleUnionPattern(((SPARQLUnionPattern) i), fedAccessMgr)) {
+						if (groupTranslation.contains(j)) {
+							partialTranslation.add(j);
+						}
+					}
+					groupTranslation = partialTranslation;
+				}
+			} else if (i instanceof SPARQLGroupPattern) {
+				if (groupTranslation == null) {
+					groupTranslation = handleGroupPattern(((SPARQLGroupPattern) i), fedAccessMgr);
+				} else {
+					final Set<Triple> partialTranslation = new HashSet<>();
+					for(final Triple j : handleGroupPattern(((SPARQLGroupPattern) i), fedAccessMgr)) {
+						if (groupTranslation.contains(j)) {
+							partialTranslation.add(j);
+						}
+					}
+					groupTranslation = partialTranslation;
+				}
+				
+			} else if (i instanceof BGP) {
+				if (groupTranslation == null) {
+					groupTranslation = handleBGP(((BGP) i), fedAccessMgr);
+				} else {
+					final Set<Triple> partialTranslation = new HashSet<>();
+					for(final Triple j : handleBGP(((BGP) i), fedAccessMgr)) {
+						if (groupTranslation.contains(j)) {
+							partialTranslation.add(j);
+						}
+					}
+					groupTranslation = partialTranslation;
+				}
+			} else {
+				throw new FederationAccessException(i.toString(), req, fm);
+			}
+		}
+		return groupTranslation;
+	}
+	
+	protected Set<Triple> handleBGP(final BGP bgp, final FederationAccessManager fedAccessMgr) throws FederationAccessException{
+		Set<Triple> bgpTranslation = null;
+		for(final TriplePattern i : bgp.getTriplePatterns()) {
+			if (bgpTranslation == null) {
+				bgpTranslation = handleTriplePattern(i, fedAccessMgr);
+			} else {
+				final Set<Triple> partialRes = new HashSet<>();
+				for (Triple j : handleTriplePattern(i, fedAccessMgr)) {
+					if (bgpTranslation.contains(j)) {
+						partialRes.add(j);
+					}
+				}
+				bgpTranslation = partialRes;
+			}
+		}
+		return bgpTranslation;
+	}
+	
+	protected Set<Triple> translateResultTriple(final Triple t, final TPFRequest q){
```

Comment:
Are you using the `TPFRequest` for something else than getting the triple pattern? If not, please consider passing the triple pattern as parameter instead.